docs/extending-turnilo.md

```diff
@@ -0,0 +1,133 @@
+
+```yaml
+druidRequestDecorator: 
+    path: './druid-request-decorator.js'
+    options:
+        keyA: valueA
+        keyB:
+          - firstElement
+          - secondElement
+```
+
+Then the contract is that your module should export a function `druidRequestDecorator` that has to return a decorator.
+ 
+A decorator is a function that gets called on every request and receives a Druid query and may return an object with the
+key `headers` where you can set whatever headers you want.
+
+Here is an example decorator:
+
+```javascript
+exports.version = 1;
+
+exports.druidRequestDecoratorFactory = function (logger, params) {
+  const options = params.options;
+  const username = options.username;
+  const password = options.password;
+
+  const auth = "Basic " + Buffer.from(`${username}:${password}`).toString("base64");
+
+  return function () {
+    return {
+      headers: {
+        "Authorization": auth
+      },
+    };
+  };
+};
+```
+
+You can find this example with additional comments and example config in the [./example](./example/request-decoration) folder.
+
+This would result in all Druid requests being tagged as:
+
+![decoration example](./example/request-decoration/result.png)
+
+## Query decorator
+
+In the data cube config add `queryDecorator` field with key `path` pointing to javascript file. 
+This file should export function named `decorator`. 
+This function will be called before every Plywood query is sent to Druid from Turnilo.
+Function is called with four arguments:
+* Plywood query
+* Request object
+* Decorator options
+* Plywood library instance
+
+Decorator function should return valid Plywood expression.
+
+```javascript
+exports.decorator = function (expression, request, options, plywood) {
+  const userId = request.headers["x-user-id"]; // get userId from header, you need to set this value before Turnilo
+  const userColumnName = options.userColumnName; // get value from options, defined in config
+  const filterClause = plywood.$(userColumnName).in([userId]); // show only rows where `userColumnName` is equal to current user id.
+  return expression.substitute(e => {
+    if (e instanceof plywood.RefExpression && e.name === "main") { // filter all main expression references
+      e.filter(filterClause);
+    }
+    return null;
+  });
+}
+```
+
+And needed configuration:
+
+```yaml
+...
+dataCubes:
+    - name: cube
+      queryDecorator:
+        path: ./decorator.js
+        options:
+          userColumnName: "user_id"
+```
+
+## Plugins
+
+Most powerful way to extend turnilo are plugins. They are defined at top level in config and apply for whole Turnilo application.
+You need to add your plugin as entry under `plugins` field. 
+Plugin need to have two fields:
+    - `name` - name for debug purposes
+    - `path` - path to the js file
+And can have additional field `settings`. Content of this field would be passed to plugin so it is good place to plugin parameters.
```

Comment:
```suggestion
It can define additional field `settings`. Content of this field would be passed to plugin so it is good place for additional parameters.
```